src/data.jl

```diff
@@ -1292,7 +1297,13 @@ jlconvert(::ReadRepresentation{Core.TypeofBottom,nothing}, f::JLDFile, ptr::Ptr,
                 push!(args, :($fsym = jlconvert($rr, f, ptr+$offset, NULL_REFERENCE)::$rtype))
             else
                 ttype = T.types[i]
-                push!(args, :($fsym = convert($ttype, jlconvert($rr, f, ptr+$offset, NULL_REFERENCE)::$rtype)::$ttype))
+                                fni = QuoteNode(fn[i])
+                if T.mutable
+                    push!(args, :(setfield!(obj, $fni,
```

Comment:
I suspect we should use `setfield!` unconditionnally. The current code uses `setproperty!` only because `=` now lowers to it, but when it was written it called `setfield!`. Immutable types can be restored just by setting their fields, I don't think it makes sense to use `setproperty!`. Anyway I doubt any package relies on this at this point.

That said, I'm not an expert of this package at all.

Reply:
I agree, but I am also not an expert and do not want to break something (e.g. some case I have not encountered) and what I changed is what I think is correct and can test. For sure - if someone knows the package it would be better to redesign the whole logic of this function.

Reply:
OK. If someone does not come up with a good solution in the mean time I will try to redesign the whole function in a few days.

Reply:
But does it really need to be redesigned? If we were on Julia 0.6, we could replace `=` with `setfield!` without changing anything to the actual behavior. If the function had been written that way originally, nothing would have to be changed. The use of `=` appears to be purely accidental to me. So I'm tempted to use `setfield!` everywhere, which is probably more correct since these have never been supposed to call custom user-defined functions via `setproperty!`.

Reply:
`fsym` originally is stored in `fsyms` vector that is used twice in this function and I did not investigate in detail for what purpose (lines 1270 and 1288), so if we change it it would be better to understand the consequences.

As for line 1275 I am not clear how to write a test case for this. If you know how to do it I would change this line and add the test.

Reply:
> fsym originally is stored in fsyms vector that is used twice in this function and I did not investigate in detail for what purpose (lines 1270 and 1288), so if we change it it would be better to understand the consequences.

Yeah, I guess `Expr(:., :obj, QuoteNode(fn[i]))` should be replaced with a call to `getfield` (but I'm very bad at metaprogramming, so don't ask me for the exact expression). IIUC `fsyms` is used to build objects while leaving the final argument unspecified (and therefore uninitialized).

> As for line 1275 I am not clear how to write a test case for this. If you know how to do it I would change this line and add the test.

AFAICT this would already fail in Julia 0.6 for immutable types, and that line hasn't been touched for two years. So `setfield!` should be safe.